engine/classes/Elgg/Search/SearchService.php

```diff
@@ -75,66 +75,110 @@ public function __construct(Config $config, PluginHooksService $hooks, Database
 	 * @see    elgg_get_entities()
 	 */
 	public function search(array $options = []) {
-
-		$search_type = elgg_extract('search_type', $options, 'entities', false);
-		if (!$search_type) {
-			return false;
-		}
-
-		$options = $this->hooks->trigger('search:params', $search_type, $options, $options);
-
-		$query = elgg_extract('query', $options);
-		if (empty($query)) {
+		$query_parts = elgg_extract('query_parts', $options);
+		if (empty($query_parts)) {
 			return false;
 		}
 
 		$options = $this->prepareSearchOptions($options);
 
-		$entity_type = elgg_extract('type', $options);
+		$entity_type = elgg_extract('type', $options, 'all', false);
 		$entity_subtype = elgg_extract('subtype', $options);
 		$search_type = elgg_extract('search_type', $options, 'entities');
 
-		if ($entity_type) {
-			$entity_type = elgg_extract('type', $options);
-			if (!in_array($entity_type, Config::getEntityTypes())) {
-				throw new InvalidParameterException("'$entity_type' is not a valid entity type");
-			}
-			
-			$options = $this->hooks->trigger('search:options', $entity_type, $options, $options);
-			if ($entity_subtype) {
-				$options = $this->hooks->trigger('search:options', "$entity_type:$entity_subtype", $options, $options);
-			}
+		if ($entity_type !== 'all' && !in_array($entity_type, Config::getEntityTypes())) {
+			throw new InvalidParameterException("'$entity_type' is not a valid entity type");
+		}
+
+		$options = $this->hooks->trigger('search:options', $entity_type, $options, $options);
+		if ($entity_subtype) {
+			$options = $this->hooks->trigger('search:options', "$entity_type:$entity_subtype", $options, $options);
 		}
 
 		$options = $this->hooks->trigger('search:options', $search_type, $options, $options);
 
+		if ($this->hooks->hasHandler('search:results', $search_type)) {
```

Comment:
could you change this to the regular pattern where you just fire the hook... and only if NULL do elgg_get_entities? That pattern is need so in hooks you can conditionally return the results...

Reply:
Why wouldn't you just call ege in your hook? Just wondering

Reply:
because then i would assume what the default behaviour is... i know of multiple plugins that can be active at the same time that could take over the search results (or modify them) but determine so (conditionally) in the results hook...